source/Calamari/Kubernetes/ResourceStatus/ResourceStatusChecker.cs

```diff
@@ -30,75 +41,144 @@ public class ResourceStatusChecker : IResourceStatusChecker
         }
-        
-        public bool CheckStatusUntilCompletionOrTimeout(IEnumerable<ResourceIdentifier> resourceIdentifiers, 
+
+        public async Task<bool> CheckStatusUntilCompletionOrTimeout(IKubectl kubectl,
+            IEnumerable<ResourceIdentifier> initialResources,
+            ITimer timer,
+            Options options)
+        {
+            StartCheckingStatus(kubectl, initialResources, timer, options);
+            return await CompleteCheckingStatus();
+        }
+
+        public void StartCheckingStatus(
+            IKubectl kubectl,
+            IEnumerable<ResourceIdentifier> initialResources,
             ITimer timer,
-            Kubectl kubectl,
             Options options)
         {
-            List<Resource> definedResourceStatuses;
-            var resourceStatuses = new Dictionary<string, Resource>();
-            var deploymentStatus = DeploymentStatus.InProgress;
+            if (isCheckingStatus)
+                throw new InvalidOperationException(
+                    "method CheckStatusUntilCompletionOrTimeout must only be called once.");
+
+            isCheckingStatus = true;
+
+            AddResourcesInternal(initialResources);
+
+            this.timer = timer;
+
+            if (!kubectl.TrySetKubectl())
+            {
+                throw new Exception("Unable to set KubeCtl");
+            }
+
             var checkCount = 0;
-            var definedResources = resourceIdentifiers.ToList();
-            
+
+            waitingForMoreResources = true;
             timer.Start();
-            
-            do
+            statusCheckTask = Task.Run(() =>
             {
-                var newDefinedResourceStatuses = resourceRetriever
-                    .GetAllOwnedResources(definedResources, kubectl, options)
-                    .ToList();
-                var newResourceStatuses = newDefinedResourceStatuses
-                    .SelectMany(IterateResourceTree)
-                    .ToDictionary(resource => resource.Uid, resource => resource);
-
-                var newDeploymentStatus = GetDeploymentStatus(newDefinedResourceStatuses, definedResources);
-
-                reporter.ReportUpdatedResources(resourceStatuses, newResourceStatuses, ++checkCount);
-
-                resourceStatuses = newResourceStatuses;
-                definedResourceStatuses = newDefinedResourceStatuses;
-                deploymentStatus = newDeploymentStatus;
-                
-                timer.WaitForInterval();
-            }
-            while (!timer.HasCompleted() && deploymentStatus == DeploymentStatus.InProgress);
+                var result = new ResourceStatusTaskResult();
+                do
+                {
+                    ResourceIdentifier[] definedResources;
+                    bool stillWaitingForMoreResources;
+                    lock (resourceLock)
+                    {
+                        stillWaitingForMoreResources = waitingForMoreResources;
```

Comment:
Set the field value of `waitingForMoreResources` to a local value here so that we ensure that we get at least 1 pass of checks before it completes as otherwise it may miss some resources if the timing is just right.